File: src/main/java/com/verlumen/tradestream/pipeline/PipelineModule.java
package com.verlumen.tradestream.pipeline;

import static com.google.protobuf.util.Timestamps.fromMillis;

import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableList;
import com.google.inject.AbstractModule;
import com.google.inject.Provider;
import com.google.inject.Provides;
import com.google.inject.Singleton;
import com.verlumen.tradestream.backtesting.BacktestingModule;
import com.verlumen.tradestream.discovery.ProdDiscoveryModule;
import com.verlumen.tradestream.execution.RunMode;
import com.verlumen.tradestream.http.HttpModule;
import com.verlumen.tradestream.influxdb.InfluxDbModule;
import com.verlumen.tradestream.instruments.InstrumentsModule;
import com.verlumen.tradestream.kafka.KafkaModule;
import com.verlumen.tradestream.marketdata.CandleSource;
import com.verlumen.tradestream.marketdata.DryRunTradeSource;
import com.verlumen.tradestream.marketdata.FillForwardCandles;
import com.verlumen.tradestream.marketdata.MarketDataModule;
import com.verlumen.tradestream.marketdata.TiingoCryptoCandleSource;
import com.verlumen.tradestream.marketdata.Trade;
import com.verlumen.tradestream.marketdata.TradeBackedCandleSource;
import com.verlumen.tradestream.marketdata.TradeSource;
import com.verlumen.tradestream.marketdata.TradeToCandle;
import com.verlumen.tradestream.postgres.PostgresModule;
import com.verlumen.tradestream.signals.SignalsModule;
import com.verlumen.tradestream.strategies.StrategiesModule;
import com.verlumen.tradestream.ta4j.Ta4jModule;
import org.joda.time.Duration;

@AutoValue
abstract class PipelineModule extends AbstractModule {
  static PipelineModule create(
      String bootstrapServers,
      int candleDurationMinutes,
      String coinMarketCapApiKey,
      String exchangeName,
      int maxForwardIntervals,
      RunMode runMode,
      String signalTopic,
      int topCurrencyCount,
      String tiingoApiKey,
      String influxDbUrl,
      String influxDbToken,
      String influxDbOrg,
      String influxDbBucket) {
    return new AutoValue_PipelineModule(
        bootstrapServers,
        Duration.standardMinutes(candleDurationMinutes),
        coinMarketCapApiKey,
        exchangeName,
        maxForwardIntervals,
        runMode,
        signalTopic,
        topCurrencyCount,
        tiingoApiKey,
        influxDbUrl,
        influxDbToken,
        influxDbOrg,
        influxDbBucket);
  }

  abstract String bootstrapServers();

  abstract Duration candleDuration();

  abstract String coinMarketCapApiKey();

  abstract String exchangeName();

  abstract int maxForwardIntervals();

  abstract RunMode runMode();

  abstract String signalTopic();

  abstract int topCurrencyCount();

  abstract String tiingoApiKey();

  abstract String influxDbUrl();

  abstract String influxDbToken();

  abstract String influxDbOrg();

  abstract String influxDbBucket();

  @Override
  protected void configure() {
    install(new BacktestingModule());
    install(new ProdDiscoveryModule());
    install(HttpModule.create());
    install(new InfluxDbModule(influxDbUrl(), influxDbToken(), influxDbOrg(), influxDbBucket()));
    install(InstrumentsModule.create(runMode(), coinMarketCapApiKey(), topCurrencyCount()));
    install(KafkaModule.create(bootstrapServers()));
    install(MarketDataModule.create()); // No parameters needed anymore
    install(new PostgresModule());
    install(SignalsModule.create(signalTopic()));
    install(new StrategiesModule());
    install(Ta4jModule.create());
  }

  @Provides
  @Singleton
  CandleSource provideCandleSource(
      Provider<TradeBackedCandleSource> tradeBackedCandleSource,
      TiingoCryptoCandleSource.Factory tiingoCryptoCandleSourceFactory) {
    switch (runMode()) {
      case DRY:
        return tradeBackedCandleSource.get();
      case WET:
        return tiingoCryptoCandleSourceFactory.create(candleDuration(), tiingoApiKey());
      default:
        throw new UnsupportedOperationException("Unsupported RunMode: " + runMode());
    }
  }

  @Provides
  @Singleton
  TradeSource provideTradeSource() {
    switch (runMode()) {
      case DRY:
        return DryRunTradeSource.create(
            ImmutableList.of(
                Trade.newBuilder()
                    .setExchange(exchangeName())
                    .setCurrencyPair("DRY/RUN")
                    .setTradeId("trade-123")
                    .setTimestamp(fromMillis(1259999L))
                    .setPrice(50000.0)
                    .setVolume(0.1)
                    .build()));
      default:
        throw new UnsupportedOperationException("Unsupported RunMode: " + runMode());
    }
  }

  @Provides
  FillForwardCandles provideFillForwardCandles(FillForwardCandles.Factory factory) {
    return factory.create(candleDuration(), maxForwardIntervals());
  }

  @Provides
  TimingConfig provideTimingConfig() {
    return TimingConfig.create();
  }

  @Provides
  TradeToCandle provideTradeToCandle(TradeToCandle.Factory factory) {
    return factory.create(candleDuration());
  }
}

File: src/main/java/com/verlumen/tradestream/pipeline/PipelineModule.kt
@file:JvmName("PipelineModuleUtility")
package com.verlumen.tradestream.pipeline

import com.google.protobuf.util.Timestamps.fromMillis
import com.google.inject.AbstractModule
import com.google.inject.Provides
import com.verlumen.tradestream.backtesting.BacktestingModule
import com.verlumen.tradestream.execution.ExecutionModule
import com.verlumen.tradestream.execution.RunMode
import com.verlumen.tradestream.kafka.DryRunKafkaReadTransform
import com.verlumen.tradestream.kafka.KafkaModule
import com.verlumen.tradestream.kafka.KafkaReadTransform
import com.verlumen.tradestream.marketdata.Trade
import com.verlumen.tradestream.signals.SignalsModule
import com.verlumen.tradestream.strategies.StrategiesModule
import com.verlumen.tradestream.ta4j.Ta4jModule
import org.apache.kafka.common.serialization.ByteArrayDeserializer
import org.apache.kafka.common.serialization.StringDeserializer

// In Java, this would be equivalent to a package-private class with static factory method
class PipelineModule private constructor(private val config: PipelineConfig) : AbstractModule() {
    private companion object {
        private val DRY_RUN_TRADE = Trade.newBuilder()
            .setExchange("FakeExhange")
            .setCurrencyPair("DRY/RUN")
            .setTradeId("trade-123")
            .setTimestamp(fromMillis(1234567))
            .setPrice(50000.0)
            .setVolume(0.1)
            .build()
    }

    override fun configure() {
        install(BacktestingModule.create())
        install(ExecutionModule.create(config.runMode()))
        install(KafkaModule.create(config.bootstrapServers()))
        install(SignalsModule.create(config.signalTopic()))
        install(StrategiesModule.create())
        install(Ta4jModule.create())
    }

    @Provides
    fun provideKafkaReadTransform(factory: KafkaReadTransform.Factory, runMode: RunMode): KafkaReadTransform<String, ByteArray> =
        if (runMode == RunMode.DRY) {
            DryRunKafkaReadTransform.builder<String, ByteArray>()
                .setBootstrapServers(config.bootstrapServers())
                .setTopic(config.tradeTopic())
                .setKeyDeserializerClass(StringDeserializer::class.java)
                .setValueDeserializerClass(ByteArrayDeserializer::class.java)
                .setDefaultValue(DRY_RUN_TRADE.toByteArray())
                .build()
        } else {
            factory.create(
                config.tradeTopic(),
                StringDeserializer::class.java,
                ByteArrayDeserializer::class.java
            )
        }

    @Provides
    fun providePipelineConfig(): PipelineConfig = config
}

// This is the Java-accessible static factory method
@JvmName("create")
fun createPipelineModule(config: PipelineConfig): PipelineModule = 
    PipelineModule::class.java.getDeclaredConstructor(PipelineConfig::class.java).apply {
        isAccessible = true 
    }.newInstance(config)
